File: VB/CustomAppointmentForm.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Text
Imports DevExpress.XtraScheduler
Imports DevExpress.XtraScheduler.UI
Imports System.Windows.Forms

Namespace SchedulerMultiResAppointments
	Partial Public Class CustomAppointmentForm
		Inherits DevExpress.XtraEditors.XtraForm
		Private control As SchedulerControl
		Private apt As Appointment
		Private openRecurrenceForm As Boolean = False
		Private suspendUpdateCount As Integer

		' The CustomAppointmentFormController class is inherited from
		' the AppointmentFormController to add custom properties.
		' See its declaration below.
		Private controller As CustomAppointmentFormController

		Protected ReadOnly Property Appointments() As AppointmentStorage
			Get
				Return control.Storage.Appointments
			End Get
		End Property

		Protected ReadOnly Property IsUpdateSuspended() As Boolean
			Get
				Return suspendUpdateCount > 0
			End Get
		End Property

		Public Sub New(ByVal control As SchedulerControl, ByVal apt As Appointment, ByVal openRecurrenceForm As Boolean)
			Me.openRecurrenceForm = openRecurrenceForm
			Me.controller = New CustomAppointmentFormController(control, apt)
			Me.apt = apt
			Me.control = control

			' Required for Windows Form Designer support
			SuspendUpdate()
			InitializeComponent()
			ResumeUpdate()
			UpdateForm()

			' TODO: Add any constructor code after InitializeComponent call
			Me.edResources.SchedulerControl = control
		End Sub

		#Region "Recurrence"
		Private Sub MyAppointmentEditForm_Activated(ByVal sender As Object, ByVal e As System.EventArgs) Handles MyBase.Activated
			' Required to show the recurrence form.
			If openRecurrenceForm Then
				openRecurrenceForm = False
				OnRecurrenceButton()
			End If
		End Sub
		Private Sub btnRecurrence_Click(ByVal sender As Object, ByVal e As System.EventArgs) Handles btnRecurrence.Click
			OnRecurrenceButton()
		End Sub

		Private Sub OnRecurrenceButton()
			ShowRecurrenceForm()
		End Sub

		Private Sub ShowRecurrenceForm()
			If (Not control.SupportsRecurrence) Then
				Return
			End If

			' Prepare to edit the appointment's recurrence.
			Dim editedAptCopy As Appointment = controller.EditedAppointmentCopy
			Dim editedPattern As Appointment = controller.EditedPattern
			Dim patternCopy As Appointment = controller.PrepareToRecurrenceEdit()

			Dim dlg As New AppointmentRecurrenceForm(patternCopy, control.OptionsView.FirstDayOfWeek, controller)

			' Required for skin support.
			dlg.LookAndFeel.ParentLookAndFeel = Me.LookAndFeel.ParentLookAndFeel

			Dim result As DialogResult = dlg.ShowDialog(Me)
			dlg.Dispose()

			If result = System.Windows.Forms.DialogResult.Abort Then
				controller.RemoveRecurrence()
			Else
				If result = System.Windows.Forms.DialogResult.OK Then
					controller.ApplyRecurrence(patternCopy)
					If controller.EditedAppointmentCopy IsNot editedAptCopy Then
						UpdateForm()
					End If
				End If
			End If
			UpdateIntervalControls()
		End Sub

		#End Region

		#Region "Form control events"

		Private Sub dtStart_EditValueChanged(ByVal sender As Object, ByVal e As System.EventArgs) Handles dtStart.EditValueChanged
			If (Not IsUpdateSuspended) Then
				controller.DisplayStart = dtStart.DateTime.Date + timeStart.Time.TimeOfDay
			End If
			UpdateIntervalControls()
		End Sub

		Private Sub timeStart_EditValueChanged(ByVal sender As Object, ByVal e As System.EventArgs) Handles timeStart.EditValueChanged
			If (Not IsUpdateSuspended) Then
				controller.DisplayStart = dtStart.DateTime.Date + timeStart.Time.TimeOfDay
			End If
			UpdateIntervalControls()
		End Sub
		Private Sub timeEnd_EditValueChanged(ByVal sender As Object, ByVal e As System.EventArgs) Handles timeEnd.EditValueChanged
			If IsUpdateSuspended Then
				Return
			End If
			If IsIntervalValid() Then
				controller.DisplayEnd = dtEnd.DateTime.Date + timeEnd.Time.TimeOfDay
			Else
				timeEnd.EditValue = New DateTime(controller.DisplayEnd.TimeOfDay.Ticks)
			End If

		End Sub
		Private Sub dtEnd_EditValueChanged(ByVal sender As Object, ByVal e As System.EventArgs) Handles dtEnd.EditValueChanged
			If IsUpdateSuspended Then
				Return
			End If
			If IsIntervalValid() Then
				controller.DisplayEnd = dtEnd.DateTime.Date + timeEnd.Time.TimeOfDay
			Else
				dtEnd.EditValue = controller.DisplayEnd.Date
			End If
		End Sub
		Private Function IsIntervalValid() As Boolean
			Dim start As DateTime = dtStart.DateTime + timeStart.Time.TimeOfDay
			Dim [end] As DateTime = dtEnd.DateTime + timeEnd.Time.TimeOfDay
			Return [end] >= start
		End Function

		Private Sub checkAllDay_CheckedChanged(ByVal sender As Object, ByVal e As System.EventArgs) Handles checkAllDay.CheckedChanged
			controller.AllDay = Me.checkAllDay.Checked
			If (Not IsUpdateSuspended) Then
				UpdateAppointmentStatus()
			End If

			UpdateIntervalControls()
		End Sub
		#End Region

		#Region "Updating Form"
		Protected Sub SuspendUpdate()
			suspendUpdateCount += 1
		End Sub
		Protected Sub ResumeUpdate()
			If suspendUpdateCount > 0 Then
				suspendUpdateCount -= 1
			End If
		End Sub

		Private Sub UpdateForm()
			SuspendUpdate()
			Try
				txSubject.Text = controller.Subject
				edStatus.Status = Appointments.Statuses(controller.StatusId)
				edLabel.Label = Appointments.Labels(controller.LabelId)

				dtStart.DateTime = controller.DisplayStart.Date
				dtEnd.DateTime = controller.DisplayEnd.Date

				timeStart.Time = New DateTime(controller.DisplayStart.TimeOfDay.Ticks)
				timeEnd.Time = New DateTime(controller.DisplayEnd.TimeOfDay.Ticks)
				checkAllDay.Checked = controller.AllDay

				edStatus.Storage = control.Storage
				edLabel.Storage = control.Storage

				txPrice.Text = controller.Price.ToString()

				' Update resource selector
				edResources.ResourceIds.Clear()
				edResources.ResourceIds.Add(controller.ResourceId)
			Finally
				ResumeUpdate()
			End Try
			UpdateIntervalControls()
		End Sub

		Protected Overridable Sub UpdateIntervalControls()
			If IsUpdateSuspended Then
				Return
			End If

			SuspendUpdate()

			Try
				dtStart.EditValue = controller.DisplayStart.Date
				dtEnd.EditValue = controller.DisplayEnd.Date
				timeStart.EditValue = New DateTime(controller.DisplayStart.TimeOfDay.Ticks)
				timeEnd.EditValue = New DateTime(controller.DisplayEnd.TimeOfDay.Ticks)

				timeStart.Visible = Not controller.AllDay
				timeEnd.Visible = Not controller.AllDay
				timeStart.Enabled = Not controller.AllDay
				timeEnd.Enabled = Not controller.AllDay
			Finally
				ResumeUpdate()
			End Try
		End Sub

		Private Sub UpdateAppointmentStatus()
			Dim currentStatus As AppointmentStatus = edStatus.Status
			Dim newStatus As AppointmentStatus = controller.UpdateAppointmentStatus(currentStatus)

			If newStatus IsNot currentStatus Then
				edStatus.Status = newStatus
			End If
		End Sub

		#End Region

		#Region "Save changes"
		Private Sub btnOK_Click(ByVal sender As Object, ByVal e As System.EventArgs) Handles btnOK.Click
			' Required to check the appointment for conflicts.
			If (Not controller.IsConflictResolved()) Then
				Return
			End If

			controller.Subject = txSubject.Text
			controller.SetStatus(edStatus.Status)
			controller.SetLabel(edLabel.Label)
			controller.AllDay = Me.checkAllDay.Checked
			controller.DisplayStart = Me.dtStart.DateTime.Date + Me.timeStart.Time.TimeOfDay
			controller.DisplayEnd = Me.dtEnd.DateTime.Date + Me.timeEnd.Time.TimeOfDay
			controller.Price = Convert.ToDecimal(txPrice.Text)

			controller.ResourceId = edResources.ResourceIds(0)

			controller.ApplyChanges()

			For Each item As Object In edResources.ResourceIds
				If item.Equals(controller.ResourceId) Then
					Continue For
				End If

				Dim apt As Appointment = controller.EditedAppointmentCopy.Copy()

				apt.ResourceId = item
				control.Storage.Appointments.Add(apt)
			Next item
		End Sub
		#End Region

		#Region "CustomAppointmentFormController"
		Public Class CustomAppointmentFormController
			Inherits AppointmentFormController
			Public Property Price() As Decimal
				Get
					Dim val As Object = EditedAppointmentCopy.CustomFields("Field1")

					If val Is Nothing OrElse val Is DBNull.Value Then
						Return 0
					Else
						Return CDec(val)
					End If
				End Get
				Set(ByVal value As Decimal)
					EditedAppointmentCopy.CustomFields("Field1") = value
				End Set
			End Property

			Private Property SourcePrice() As Decimal
				Get
					Return CDec(SourceAppointment.CustomFields("Field1"))
				End Get
				Set(ByVal value As Decimal)
					SourceAppointment.CustomFields("Field1") = value
				End Set
			End Property

			Public Sub New(ByVal control As SchedulerControl, ByVal apt As Appointment)
				MyBase.New(control, apt)
			End Sub

			Public Overrides Function IsAppointmentChanged() As Boolean
				If MyBase.IsAppointmentChanged() Then
					Return True
				End If
				Return SourcePrice <> Price
			End Function

			Protected Overrides Sub ApplyCustomFieldsValues()
				SourcePrice = Price
			End Sub
		End Class
		#End Region
	End Class

End Namespace

File: CS/CustomAppointmentForm.cs
using System;
using System.Collections.Generic;
using System.Text;
using DevExpress.XtraScheduler;
using DevExpress.XtraScheduler.UI;
using System.Windows.Forms;

namespace SchedulerMultiResAppointments {
    public partial class CustomAppointmentForm : DevExpress.XtraEditors.XtraForm {
        SchedulerControl control;
        Appointment apt;
        bool openRecurrenceForm = false;
        int suspendUpdateCount;

        // The CustomAppointmentFormController class is inherited from
        // the AppointmentFormController to add custom properties.
        // See its declaration below.
        CustomAppointmentFormController controller;

        protected IAppointmentStorage Appointments {
            get { return control.DataStorage.Appointments; }
        }

        protected bool IsUpdateSuspended { 
            get { return suspendUpdateCount > 0; } 
        }

        public CustomAppointmentForm(SchedulerControl control, Appointment apt, bool openRecurrenceForm) {
            this.openRecurrenceForm = openRecurrenceForm;
            this.controller = new CustomAppointmentFormController(control, apt);
            this.apt = apt;
            this.control = control;
            
            // Required for Windows Form Designer support
            SuspendUpdate();
            InitializeComponent();
            ResumeUpdate();
            UpdateForm();
            
            // TODO: Add any constructor code after InitializeComponent call
            this.edResources.SchedulerControl = control;
        }

        #region Recurrence
        private void MyAppointmentEditForm_Activated(object sender, System.EventArgs e) {
            // Required to show the recurrence form.
            if (openRecurrenceForm) {
                openRecurrenceForm = false;
                OnRecurrenceButton();
            }
        }
        private void btnRecurrence_Click(object sender, System.EventArgs e) {
            OnRecurrenceButton();
        }

        void OnRecurrenceButton() {
            ShowRecurrenceForm();
        }

        void ShowRecurrenceForm() {
            if (!control.SupportsRecurrence)
                return;

            // Prepare to edit the appointment's recurrence.
            Appointment editedAptCopy = controller.EditedAppointmentCopy;
            Appointment editedPattern = controller.EditedPattern;
            Appointment patternCopy = controller.PrepareToRecurrenceEdit();

            AppointmentRecurrenceForm dlg = new AppointmentRecurrenceForm(patternCopy, control.OptionsView.FirstDayOfWeek, controller);

            // Required for skin support.
            dlg.LookAndFeel.ParentLookAndFeel = this.LookAndFeel.ParentLookAndFeel;

            DialogResult result = dlg.ShowDialog(this);
            dlg.Dispose();

            if (result == DialogResult.Abort)
                controller.RemoveRecurrence();
            else
                if (result == DialogResult.OK) {
                    controller.ApplyRecurrence(patternCopy);
                    if (controller.EditedAppointmentCopy != editedAptCopy)
                        UpdateForm();
                }
            UpdateIntervalControls();
        }

        #endregion

        #region Form control events

        private void dtStart_EditValueChanged(object sender, System.EventArgs e) {
            if (!IsUpdateSuspended)
                controller.DisplayStart = dtStart.DateTime.Date + timeStart.Time.TimeOfDay;
            UpdateIntervalControls();
        }

        private void timeStart_EditValueChanged(object sender, System.EventArgs e) {
            if (!IsUpdateSuspended)
                controller.DisplayStart = dtStart.DateTime.Date + timeStart.Time.TimeOfDay;
            UpdateIntervalControls();
        }
        private void timeEnd_EditValueChanged(object sender, System.EventArgs e) {
            if (IsUpdateSuspended) return;
            if (IsIntervalValid())
                controller.DisplayEnd = dtEnd.DateTime.Date + timeEnd.Time.TimeOfDay;
            else
                timeEnd.EditValue = new DateTime(controller.DisplayEnd.TimeOfDay.Ticks);

        }
        private void dtEnd_EditValueChanged(object sender, System.EventArgs e) {
            if (IsUpdateSuspended) return;
            if (IsIntervalValid())
                controller.DisplayEnd = dtEnd.DateTime.Date + timeEnd.Time.TimeOfDay;
            else
                dtEnd.EditValue = controller.DisplayEnd.Date;
        }
        bool IsIntervalValid() {
            DateTime start = dtStart.DateTime + timeStart.Time.TimeOfDay;
            DateTime end = dtEnd.DateTime + timeEnd.Time.TimeOfDay;
            return end >= start;
        }

        private void checkAllDay_CheckedChanged(object sender, System.EventArgs e) {
            controller.AllDay = this.checkAllDay.Checked;
            if (!IsUpdateSuspended)
                UpdateAppointmentStatus();

            UpdateIntervalControls();
        }
        #endregion

        #region Updating Form
        protected void SuspendUpdate() {
            suspendUpdateCount++;
        }
        protected void ResumeUpdate() {
            if (suspendUpdateCount > 0)
                suspendUpdateCount--;
        }

        void UpdateForm() {
            SuspendUpdate();
            try {
                txSubject.Text = controller.Subject;
                edStatus.AppointmentStatus = Appointments.Statuses.GetById(controller.StatusKey);
                edLabel.AppointmentLabel = Appointments.Labels.GetById(controller.LabelKey);

                dtStart.DateTime = controller.DisplayStart.Date;
                dtEnd.DateTime = controller.DisplayEnd.Date;

                timeStart.Time = new DateTime(controller.DisplayStart.TimeOfDay.Ticks);
                timeEnd.Time = new DateTime(controller.DisplayEnd.TimeOfDay.Ticks);
                checkAllDay.Checked = controller.AllDay;

                edStatus.Storage = control.Storage;
                edLabel.Storage = control.Storage;

                txPrice.Text = controller.Price.ToString();

                // Update resource selector
                edResources.ResourceIds.Clear();
                edResources.ResourceIds.Add(controller.ResourceId);
            }
            finally {
                ResumeUpdate();
            }
            UpdateIntervalControls();
        }

        protected virtual void UpdateIntervalControls() {
            if (IsUpdateSuspended)
                return;

            SuspendUpdate();
            
            try {
                dtStart.EditValue = controller.DisplayStart.Date;
                dtEnd.EditValue = controller.DisplayEnd.Date;
                timeStart.EditValue = new DateTime(controller.DisplayStart.TimeOfDay.Ticks);
                timeEnd.EditValue = new DateTime(controller.DisplayEnd.TimeOfDay.Ticks);

                timeStart.Visible = !controller.AllDay;
                timeEnd.Visible = !controller.AllDay;
                timeStart.Enabled = !controller.AllDay;
                timeEnd.Enabled = !controller.AllDay;
            }
            finally {
                ResumeUpdate();
            }
        }

        void UpdateAppointmentStatus() {
            IAppointmentStatus currentStatus = edStatus.AppointmentStatus;
            IAppointmentStatus newStatus = controller.UpdateStatus(currentStatus);
            
            if (newStatus != currentStatus)
                edStatus.AppointmentStatus = newStatus;
        }

        #endregion

        #region Save changes
        private void btnOK_Click(object sender, System.EventArgs e) {
            // Required to check the appointment for conflicts.
            if (!controller.IsConflictResolved())
                return;

            controller.Subject = txSubject.Text;
            controller.SetStatus(edStatus.AppointmentStatus);
            controller.SetLabel(edLabel.AppointmentLabel);
            controller.AllDay = this.checkAllDay.Checked;
            controller.DisplayStart = this.dtStart.DateTime.Date + this.timeStart.Time.TimeOfDay;
            controller.DisplayEnd = this.dtEnd.DateTime.Date + this.timeEnd.Time.TimeOfDay;
            controller.Price = Convert.ToDecimal(txPrice.Text);

            controller.ResourceId = edResources.ResourceIds[0];

            controller.ApplyChanges();

            foreach (object item in edResources.ResourceIds) {
                if (item.Equals(controller.ResourceId))
                    continue;
                
                Appointment apt = controller.EditedAppointmentCopy.Copy();

                apt.ResourceId = item;
                control.DataStorage.Appointments.Add(apt);
            }
        }
        #endregion

        #region CustomAppointmentFormController
        public class CustomAppointmentFormController : AppointmentFormController {
            public decimal Price {
                get {
                    object val = EditedAppointmentCopy.CustomFields["Field1"];

                    if (val == null || val == DBNull.Value)
                        return 0;
                    else
                        return (decimal)val;
                }
                set { 
                    EditedAppointmentCopy.CustomFields["Field1"] = value; 
                }
            }

            decimal SourcePrice { get { return (decimal)SourceAppointment.CustomFields["Field1"]; } set { SourceAppointment.CustomFields["Field1"] = value; } }

            public CustomAppointmentFormController(SchedulerControl control, Appointment apt)
                : base(control, apt) {
            }

            public override bool IsAppointmentChanged() {
                if (base.IsAppointmentChanged())
                    return true;
                return SourcePrice != Price;
            }

            protected override void ApplyCustomFieldsValues() {
                SourcePrice = Price;
            }
        }
        #endregion
    }

}
